pennylane_qiskit/qiskit_device2.py

```diff
@@ -0,0 +1,606 @@
+        # also handle instances where wires were specified in probs, and for multiple probs measurements
+        # single_measurement = len(circuit.measurements) == 1
+        # res = (res[0], ) if single_measurement else tuple(res)
+
+        return (result.quasi_dists[0],)
+
+    def _execute_estimator(self, circuit, session):
+        # the Estimator primitive takes care of diagonalization and measurements itself,
+        # so diagonalizing gates and measurements are not included in the circuit
+        qcirc = circuit_to_qiskit(circuit, self.num_wires, diagonalize=False, measure=False)
+
+        estimator = Estimator(session=session, options=self.options)
+
+        # split into one call per measurement
+        # could technically be more efficient if there are some observables where we ask
+        # for expectation value and variance on the same observable, but spending time on
+        # that right now feels excessive
+        pauli_observables = [mp_to_pauli(mp, self.num_wires) for mp in circuit.measurements]
+        result = estimator.run([qcirc] * len(pauli_observables), pauli_observables).result()
+        self._current_job = result
+        result = self._process_estimator_job(circuit.measurements, result)
+
+        return result
+
+    @staticmethod
+    def _process_estimator_job(measurements, job_result):
+        """Estimator returns both expectation value and variance for each observable measured,
+        along with some metadata. Extract the relevant number for each measurement process and
+        return the requested results from the Estimator executions."""
+
+        expvals = job_result.values
+        variances = [res["variance"] for res in job_result.metadata]
+
+        result = []
+        for i, mp in enumerate(measurements):
+            if isinstance(mp, ExpectationMP):
+                result.append(expvals[i])
+            elif isinstance(mp, VarianceMP):
+                result.append(variances[i])
+
+        single_measurement = len(measurements) == 1
+        result = (result[0],) if single_measurement else tuple(result)
+
+        return result
+
+    def generate_samples(self, circuit=None):
+        r"""Returns the computational basis samples generated for all wires.
+
+        Note that PennyLane uses the convention :math:`|q_0,q_1,\dots,q_{N-1}\rangle` where
+        :math:`q_0` is the most significant bit.
+
+        Args:
+            circuit (int): position of the circuit in the batch.
+
+        Returns:
+             array[complex]: array of samples in the shape ``(dev.shots, dev.num_wires)``
+        """
+        counts = self._current_job.get_counts()
+        # Batch of circuits
+        if not isinstance(counts, dict):
+            counts = self._current_job.get_counts()[circuit]
+
+        samples = []
+        for key, value in counts.items():
+            for _ in range(0, value):
+                samples.append(key)
```

Comment:
might be more performant since there can be a lot of samples, but i'm not sure.
```suggestion
            samples.extend([key] * value)
```